module.go

```diff
@@ -5,15 +5,40 @@ type Scanner interface {
 	// Init runs once for this module at library init time
 	Init(flags ScanFlags) error
 
-	// InitPerSender runs once per Goroutine. A single Goroutine will scan some non-deterministics
+	// InitPerSender runs once per Goroutine. A single Goroutine will scan some non-deterministic
 	// subset of the input scan targets
 	InitPerSender(senderID int) error
 
 	// Returns the name passed at init
 	GetName() string
 
 	// Scan connects to a host. The result should be JSON-serializable
-	Scan(t ScanTarget) (interface{}, error)
+	Scan(t ScanTarget) (ScanStatus, interface{}, error)
```

Comment:
I think you _could_ get the same information with just result + error, but that pushes determining the "status" onto the caller (without really removing any work from the module, since it still needs to ensure that it is throwing sufficiently-granular errors for the caller to be able to make that determination).

If I'm recalling the convo with @zakir on Tueday correctly, SCAN_SUCCESS means (despite the comment in module.go) only that the protocol was successfully detected, but there could still have been an application-level error (though, the only example that springs to mind for MySQL is an update-to-TLS error, which is one of our non-success statuses).

The way I understood it, one big selling point of status was giving users a way to search for machines that were definitely running X, since searching for "error NOT NULL" is not an option.

For protocols like MySQL, result NOT NULL would probably work as a proxy for that broader definition of "success", but for fully TLS-wrapped protocols like HTTPS, it's conceivable that you could get some useful TLS logs before finding out that it's not HTTP underneath.

Reply:
My question was more, if we have ScanStatus, do we also need `error`? I think we're OK leaving all three for now.

Reply:
Ah -- in that case, I see the `error` as more an item for anyone directly using the zgrab API, or for logging/debugging purposes...e.g. we can't get any more specific than SCAN_APPLICATION_ERROR for a generic status, but the error object could give information about what the specific application error was (maybe "authentication failure" or "no users are allowed to connect from your IP address" etc).